### ISSUE TITLE
[bug] system lib (dl) ordering change on link

### ISSUE THREAD

**Issue by dfleury2** [conan-io/conan#11773]
### Environment Details (include every applicable attribute)
  * Operating System+version: Debian bullseye
  * Compiler+version: gcc 11
  * Conan version: 1.51
  * Python version: 3.9.2

### Steps to reproduce (Include if Applicable)
Not easy, but lot of dependencies and openssl with dl usage (using crypto)

### Logs (Executed commands with output) (Include/Attach if Applicable)
```
/usr/bin/c++  -pipe -Wall -fstack-protector-all -fstack-check -O3 -DNDEBUG  -Wl,-z,relro -Wl,--as-needed CMakeFiles/bin_beauty_application_singleton.dir/beauty_application_singleton.cpp.o -o ../../../bin/beauty_application_singleton  ../../../lib/libbeauty.a /home/dfleury/.conan/data/openssl/1.1.1q/_/_/package/fae585d7c68637b721c0c4c0d1c52174d067eda8/lib/libssl.a -ldl /home/dfleury/.conan/data/openssl/1.1.1q/_/_/package/fae585d7c68637b721c0c4c0d1c52174d067eda8/lib/libcrypto.a -lrt -lpthread -lstdc++fs 
```

Here -ldl is placed before crypto.a, so I get 
```
usr/bin/ld: /home/dfleury/.conan/data/openssl/1.1.1q/_/_/package/fae585d7c68637b721c0c4c0d1c52174d067eda8/lib/libcrypto.a(dso_dlfcn.o): in function `dlfcn_globallookup':
dso_dlfcn.c:(.text+0x13): undefined reference to `dlopen'
```

If I use conan 1.49 or conan 1.50, it works fine.
```
usr/bin/c++  -pipe -Wall -fstack-protector-all -fstack-check -O3 -DNDEBUG  -Wl,-z,relro -Wl,--as-needed CMakeFiles/bin_beauty_chat.dir/beauty_chat.cpp.o -o ../../../bin/beauty_chat  ../../../lib/libbeauty.a /home/dfleury/.conan/data/openssl/1.1.1q/_/_/package/fae585d7c68637b721c0c4c0d1c52174d067eda8/lib/libssl.a /home/dfleury/.conan/data/openssl/1.1.1q/_/_/package/fae585d7c68637b721c0c4c0d1c52174d067eda8/lib/libcrypto.a /home/dfleury/.conan/data/openssl/1.1.1q/_/_/package/fae585d7c68637b721c0c4c0d1c52174d067eda8/lib/libcrypto.a -ldl -lpthread -lrt -lstdc++fs 
```

**Comment by franramirez688** [comment]
Hi @dfleury2 

I would need more details about your recipe. If it's not easy to reproduce, could you try to simplify it as much as possible to have anything that replicates the behavior? Would it be enough with a "Hello World" example (`conan new hello/1.0 -m cmake_lib` for instance) and openssl (+ dl usage) as requirement? It's important to know the generator, the requirements, if you're injecting flags, etc.

**Comment by dfleury2** [comment]
Ok, I will try if I can isolated something that trigger this.

At least, even if the order is wrong, if I remove ` -Wl,--as-needed` for the link step,  the error disapear (but dl is still at the wrong place)

**Comment by dfleury2** [comment]
I forgot to mention but I used CMakeDeps and CMakeToolchain generator.
I a
m using OpenSSL along with Boost.
When I compare the files generated with conan 1.50, and 1.51.
The fles OpenSSL-release-x86_64-data.cmake and OpenSSL-Target-release.cmake have change a lot.

But, there is a small carriage missing now (1.51), 
![image](https://user-images.githubusercontent.com/4805384/182826018-e033ff59-6dd3-4bce-b3ec-7eda19a69623.png)
just before the comment, in the OpenSSL-release file

**Comment by lasote** [comment]
Thanks! we were wondering how you used it. 
Could you show how are you "linking" the targets in your project? I mean `find_package` + `target_link_libraries(foo OpenSSL:SSL)` or are you using a variable like `openssl_LIBRARIES` or so?

We have some ongoing patches for the 1.51 on CMakeDeps but would be great to know if we actually fixed them also for you. If you could provide the CMakeLists.txt of your consumer would help a lot. (Or a small project failing)

**Comment by dfleury2** [comment]
Ok, I found something that work, on a direct dependency.
Currently I am using `openssl::openssl`as dependency (but broken).
If I use 
```
OpenSSL::crypto
OpenSSL::SSL
```
in this exact order it works fine. If I put SSL before crypto, it's broken again, the dl lib is after ssl.a but before crypto.a
Hope this one help.

**Comment by dfleury2** [comment]
I am using some "scripts" to generate lot of boring stuff. Hard to extract it easily.
But something as simple as:
```
find_package(Boost CONFIG REQUIRED)
find_package(OpenSSL CONFIG REQUIRED)

add_library(mylib ...)
target_link_library(mylib PUBLIC Boost::headers opensssl::openssl)

add_executable(myexe ...)
target_link_library(myexe PUBLIC mylib::mylib)
```
just that I generated a variable inside conanfile.py that list all dependencies generated (*.cmake files), pass it to CMake, to call each find_package.

and I have a macro to replace add_library/target_link_libraries in a single call.

**Comment by czoido** [comment]
Thanks for reporting.

Closed by https://github.com/conan-io/conan/pull/11788 to be released soon in 1.51.1
